templates/Dockerfile.tt

```diff
@@ -53,28 +58,31 @@ FROM builder AS ci
 COPY --chown=deploy:deploy infra/bundle/config.ci ${BUNDLE_PATH}/config
 COPY --chown=deploy:deploy Gemfile* /app/
 RUN bundle install && \
-    rm -rf ~/.bundle/ ${BUNDLE_PATH}/ruby/*/cache ${BUNDLE_PATH}/ruby/*/bundler/gems/*/.git<% if uses_bootsnap? %> && \
-    bundle exec bootsnap precompile --gemfile<% end %>
+  rm -rf ~/.bundle/ ${BUNDLE_PATH}/ruby/*/cache ${BUNDLE_PATH}/ruby/*/bundler/gems/*/.git<% if uses_bootsnap? %> && \
+  bundle exec bootsnap precompile --gemfile<% end %>
 COPY --chown=deploy:deploy . /app
 
 FROM builder AS deploy-base
 COPY --chown=deploy:deploy infra/bundle/config.deploy ${BUNDLE_PATH}/config
 COPY --chown=deploy:deploy Gemfile* /app/
 RUN bundle install && \
-    rm -rf ~/.bundle/ ${BUNDLE_PATH}/ruby/*/cache ${BUNDLE_PATH}/ruby/*/bundler/gems/*/.git<% if uses_bootsnap? %> && \
-    bundle exec bootsnap precompile --gemfile<% end %>
+  rm -rf ~/.bundle/ ${BUNDLE_PATH}/ruby/*/cache ${BUNDLE_PATH}/ruby/*/bundler/gems/*/.git<% if uses_bootsnap? %> && \
+  bundle exec bootsnap precompile --gemfile<% end %>
+
+COPY --chown=deploy:deploy . /app<% if uses_node? %>
```

Comment:
I think [skip_asset_pipeline?](https://github.com/rails/rails/blob/45119752f44b4fe2c8d3f39f7576947a690772c6/railties/lib/rails/generators/app_base.rb#L390) would be more appropriate compared to `uses_node?`. In case of API only app (`--api`), asset precompilation should be skipped.

Reply:
> on a second glance, it seems to me that we should have both uses_node? and skip_asset_pipeline?.

Agreed

> What do you think?

I think your logic isn't correct for the case of API only (`skip_asset_pipeline? => true`) but uses Node (`uses_node? => true`). Because of asset pipeline is skipped, `node_modules` won't be copied in this stage. I think it should be

```
COPY --chown=deploy:deploy . /app<% if uses_node? || !skip_asset_pipeline? %>
COPY --chown=deploy:deploy --from=builder /app/node_modules /app/node_modules<% end %>
<% unless skip_asset_pipeline? %>
ARG RAILS_ENV=production
RUN --mount=type=secret,id=app-secrets,target=.env \
  bundle exec rails assets:precompile
<% end %>
```
Definitely fact check me on this.